File: packages/guardian-rn/ios/Sources/GuardianRN/DeviceAuthManager.swift
import Foundation
import LocalAuthentication

/// Provides device authentication posture queries exposed via GuardianDeviceAuth TurboModule.
///
/// Mirrors the Android DeviceAuthManager contract:
///   isPasscodeSet()       — true when any device lock credential is set
///   isBiometricEnrolled() — true when Face ID or Touch ID is enrolled
///
/// Per ADR-0023.
@objc public class DeviceAuthManager: NSObject {

    /// Returns true when the device has a passcode (PIN or alphanumeric)
    /// configured. Uses LAContext.canEvaluatePolicy(.deviceOwnerAuthentication)
    /// which returns true for both biometric-and-passcode and passcode-only
    /// configurations.
    @objc public static func isPasscodeSet() -> Bool {
        let context = LAContext()
        var error: NSError?
        // .deviceOwnerAuthentication covers passcode OR biometrics
        let canEval = context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &error)
        // LAError.passcodeNotSet is set when no passcode exists
        if let laError = error as? LAError, laError.code == .passcodeNotSet {
            return false
        }
        return canEval
    }

    /// Returns true when Face ID or Touch ID is enrolled on the device.
    @objc public static func isBiometricEnrolled() -> Bool {
        let context = LAContext()
        var error: NSError?
        let canEval = context.canEvaluatePolicy(
            .deviceOwnerAuthenticationWithBiometrics,
            error: &error,
        )
        if let laError = error as? LAError {
            // biometryNotEnrolled or biometryNotAvailable both indicate no enrollment
            if laError.code == .biometryNotEnrolled || laError.code == .biometryNotAvailable {
                return false
            }
        }
        return canEval
    }
}

File: packages/guardian-rn/android/src/main/kotlin/com/guardian/rn/DeviceAuthManager.kt
package com.guardian.rn

import android.app.KeyguardManager
import android.content.Context
import android.os.Build
import androidx.biometric.BiometricManager
import androidx.biometric.BiometricManager.Authenticators.BIOMETRIC_STRONG

/**
 * Provides device authentication posture checks exposed to the JS layer
 * via the GuardianDeviceAuth TurboModule.
 *
 * isPasscodeSet()       — true when a PIN, pattern, or password is configured
 * isBiometricEnrolled() — true when at least one strong biometric is enrolled
 *
 * Per ADR-0023.
 */
class DeviceAuthManager(private val context: Context) {

    /**
     * Returns true when the device has a secure lock screen (PIN/pattern/password).
     * Uses KeyguardManager.isDeviceSecure() on API 23+.
     */
    fun isPasscodeSet(): Boolean {
        val km = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            km.isDeviceSecure
        } else {
            @Suppress("DEPRECATION")
            km.isKeyguardSecure
        }
    }

    /**
     * Returns true when at least one strong biometric (fingerprint, face, iris)
     * is enrolled. Uses BiometricManager from androidx.biometric.
     */
    fun isBiometricEnrolled(): Boolean {
        val bm = BiometricManager.from(context)
        return bm.canAuthenticate(BIOMETRIC_STRONG) == BiometricManager.BIOMETRIC_SUCCESS
    }
}
